service/src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java

```diff
@@ -176,11 +148,108 @@ public void run() {
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        RequestMetadataUpdateHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    RequestMetadata requestMetadata = requestUpdateQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (requestMetadata != null) {
+                        MetaDbRequest existingRequest =
+                                requestService.getMetadbRequestById(requestMetadata.getRequestId());
+
+                        if (requestService.requestHasMetadataUpdates(
+                                existingRequest.getLatestRequestMetadata(), requestMetadata)) {
+                            // persist request-level metadata updates to database
+                            // existingRequest.updateRequestMetadata(request);
+                            // requestService.updateRequestMetadata(existingRequest);
+
+                            //LOG.info("Publishing Request-level Metadata updates "
+                            //        + "to CMO_REQUEST_METADATA_UPDATE");
+                            // publish request-level metadata history to CMO_REQUEST_UPDATE_TOPIC ..
+                            //messagingGateway.publish(existingRequest.getRequestId(),
+                            //        CMO_REQUEST_UPDATE_TOPIC,
+                            //        mapper.writeValueAsString(
+                            //              existingRequest.getRequestMetadataList()));
+                        } else {
+                            LOG.warn("There are no request-level metadata updates to persist: "
+                                    + requestMetadata.getRequestId());
+                        }
+                    }
+                    if (interrupted && requestUpdateQueue.isEmpty()) {
+                        break;
+                    }
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                } catch (Exception e) {
+                    LOG.error("Error during handling of request metadata update", e);
+                }
+            }
+            requestUpdateHandlerShutdownLatch.countDown();
+        }
+    }
+
+    private class SampleMetadataUpdateHandler implements Runnable {
+
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        SampleMetadataUpdateHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    SampleMetadata sampleMetadata = sampleUpdateQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (sampleMetadata != null) {
+                        MetaDbSample existingSample = sampleService.getMetaDbSampleByRequestAndIgoId(
```

Comment:
Check if existingSample is null for handling new samples

Reply:
Note: If the request is not new, any updates to the request metadata or sample metadata will published to their respective yellow channels